src/Transports.AspNetCore.SystemTextJson/GraphQLRequestDeserializer.cs

```diff
@@ -0,0 +1,134 @@
+                        .Select(ToGraphQLRequest)
+                        .ToArray();
+                    return result;
+                default:
+                    result.IsSuccessful = false;
+                    return result;
+            }
+        }
+
+        private static async ValueTask<JsonTokenType> PeekJsonTokenTypeAsync(PipeReader reader, CancellationToken cancellationToken = default)
+        {
+            // Separate method so that we can use the ref struct
+            static bool DetermineTokenType(in ReadOnlySequence<byte> buffer, out JsonTokenType jsonToken)
+            {
+                var jsonReader = new Utf8JsonReader(buffer);
+                if (jsonReader.Read())
+                {
+                    jsonToken = jsonReader.TokenType;
+                    return true;
+                }
+                jsonToken = JsonTokenType.None;
+                return false;
+            }
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var result = await reader.ReadAsync(cancellationToken);
+                var buffer = result.Buffer;
+
+                if (DetermineTokenType(buffer, out var tokenType))
+                {
+                    // Don't consume any of the buffer so we can re-parse it with the
+                    // serializer
+                    reader.AdvanceTo(buffer.Start, buffer.Start);
+                    return tokenType;
+                }
+                else
+                {
+                    // We don't have enough to read a token, keep buffering
+                    reader.AdvanceTo(buffer.Start, buffer.End);
+                }
+
+                // If there's no more data coming, then bail
+                if (result.IsCompleted)
+                {
+                    return JsonTokenType.None;
+                }
+            }
+        }
+
+        public GraphQLRequest DeserializeFromQueryString(IQueryCollection qs) => ToGraphQLRequest(new InternalGraphQLRequest
+        {
+            Query = qs.TryGetValue(GraphQLRequest.QueryKey, out var queryValues) ? queryValues[0] : null,
+            Variables = qs.TryGetValue(GraphQLRequest.VariablesKey, out var variablesValues) ? variablesValues[0].ToDictionary() : null,
+            OperationName = qs.TryGetValue(GraphQLRequest.OperationNameKey, out var operationNameValues) ? operationNameValues[0] : null
+        });
+
+        public GraphQLRequest DeserializeFromFormBody(IFormCollection fc) => ToGraphQLRequest(new InternalGraphQLRequest
+        {
+            Query = fc.TryGetValue(GraphQLRequest.QueryKey, out var queryValues) ? queryValues[0] : null,
+            Variables = fc.TryGetValue(GraphQLRequest.VariablesKey, out var variablesValue) ? variablesValue[0].ToDictionary() : null,
+            OperationName = fc.TryGetValue(GraphQLRequest.OperationNameKey, out var operationNameValues) ? operationNameValues[0] : null
+        });
```

Comment:
```c#
 public GraphQLRequest DeserializeFromQueryString(IQueryCollection qs) => new GraphQLRequest
        {
            Query = qs.TryGetValue(GraphQLRequest.QueryKey, out var queryValues) ? queryValues[0] : null,
            Inputs = qs.TryGetValue(GraphQLRequest.VariablesKey, out var variablesValues) ? variablesValues[0].ToDictionary().ToInputs() : null,
            OperationName = qs.TryGetValue(GraphQLRequest.OperationNameKey, out var operationNameValues) ? operationNameValues[0] : null
        };

        public GraphQLRequest DeserializeFromFormBody(IFormCollection fc) => new GraphQLRequest
        {
            Query = fc.TryGetValue(GraphQLRequest.QueryKey, out var queryValues) ? queryValues[0] : null,
            Inputs = fc.TryGetValue(GraphQLRequest.VariablesKey, out var variablesValue) ? variablesValue[0].ToDictionary().ToInputs() : null,
            OperationName = fc.TryGetValue(GraphQLRequest.OperationNameKey, out var operationNameValues) ? operationNameValues[0] : null
        };
```

Reply:
Even better, don't think I need to to `.ToDictionary()?.ToInputs()` or similarly `JObject.Parse(x)?.ToInputs()` as there's a `ToInputs()` off `string` already. Which then means the deserialization from querystring and formbody are essentially the same between implementations and should go back to the middleware and call into an interface method that's just `string >> Inputs`

Reply:
OK